vignettes/namespaces.Rmd

```diff
@@ -0,0 +1,288 @@
+inherit the namespace, which affects your XPath expressions: 
+
+```{r}
+d <- xml2::read_xml("<document>
+    <paragraph xmlns='http://commonmark.org/xml/1.0'>
+      <text>hello there</text>
+      <text> ello  here</text>
+    </paragraph>
+  </document>")
+xml2::xml_ns(d)
+xml2::xml_find_all(d, "//document")
+xml2::xml_find_all(d, "//text[contains(text(), 'hello')]") # does not work
+xml2::xml_find_all(d, "//d1:text[contains(text(), 'hello')]") # works; default namespace
+```
+
+When a namespace is specified with `xmlns=<URI>`, {xml2} assigns it a **default
+namespace** prefix, which is `d1`. The {xml2} documentation recommends to
+rename the namespace as soon as you read in a document and use the namespace
+object to semantically prefix your XPath expressions:
+
+```{r}
+ns <- xml2::xml_ns(d)
+ns <- xml2::xml_ns_rename(ns, d1 = "md") 
+ns
+xml2::xml_find_all(d, "//md:text[contains(text(), 'hello')]", ns)
+```
+
+You might be wondering, why don't we prefix the namespace from the start to 
+avoid needing to rename and specify the namespace? Here's an example. Let's take
+our previous example and modify the namespace attribute to have an `md` prefix:
+
+```{r}
+dc <- as.character(d)
+cat(dc <- gsub("xmlns=", "xmlns:md=", dc))
+dc <- xml2::read_xml(dc)
+xml2::xml_ns(dc)
+xml2::xml_find_all(dc, "//document")
+xml2::xml_find_all(dc, "//text[contains(text(), 'hello')]") # works!
+xml2::xml_find_all(dc, "//md:text") # prefix no longer works :(
+```
+
+Now the prefix syntax no longer works, but that's okay, because now we have the
+commonmark as our default namespace, right? Unfortunately, that's not quite the
+case. You might notice that we can access the `document` node AND the `text`
+node without a prefix even though the `text` node is in the commonmark namespace
+and the `document` node is outside of that namespace. It's because neither of
+these nodes actually have a namespace! 
+
+You can see this when we add a new node with a namespace prefix:
+
+```{r}
+pgp <- xml2::xml_find_first(dc, "//paragraph")
+xml2::xml_add_child(pgp, "md:text", "hello from the md namespace")
+xml2::xml_find_all(dc, "//md:text") # only one text element in the md namespace
+xml2::xml_find_all(dc, "//text")    # md namespace does not show up here
+```
+
+If a namespace is defined in the document with a prefix, *only nodes with that
+prefix are considered to be inside the namespace*. This becomes important when
+we want to pass our XML document through a stylesheet that expects the incoming
+nodes to have a specific namespace, which is exactly how we transform the XML
+representation of markdown back to markdown.
+
+## Commonmark
+
+The {tinkr} package streamlines the process of markdown to xml and back again.
+We use `commonmark::markdown_xml()` as a starting point to generate valid XML:
+
+
+```{r cmark}
+cat(cmk <- commonmark::markdown_xml("this is a **test**"))
+xml <- xml2::read_xml(cmk)
+xml
+```
+
+### It's default of the namespace
```

Comment:
I fear the cool title might be hard to understand for non native speakers.

Reply:
Agreed. I'll change this to "Default Namespaces"